libs/ballot-encoder/src/index.ts

```diff
@@ -214,48 +416,42 @@ export function decodeBallotFromReader(
 ): CompletedBallot {
   if (!bits.skipUint8(...Prelude)) {
     throw new Error(
-      "expected leading prelude 'V' 'X' 0b00000001 but it was not found"
+      "expected leading prelude 'V' 'X' 0b00000002 but it was not found"
     )
   }
 
-  const ballotStyleId = bits.readString()
-  const ballotStyle = getBallotStyle({ ballotStyleId, election })
-
-  if (!ballotStyle) {
-    throw new Error(
-      `ballot style with id ${JSON.stringify(
-        ballotStyleId
-      )} could not be found, expected one of: ${election.ballotStyles
-        .map((bs) => bs.id)
-        .join(', ')}`
-    )
-  }
+  const electionHash = bits.readString({
+    encoding: HexEncoding,
+    length: ELECTION_HASH_LENGTH,
+  })
 
-  const precinctId = bits.readString()
+  const {
+    ballotId = '',
+    ballotStyleId,
+    ballotType,
+    isTestMode,
+    precinctId,
+  } = decodeBallotConfigFromReader(
+    election,
+    { readLocales: false, readPageNumber: false },
+    bits
+  )
+  const ballotStyle = getBallotStyle({ ballotStyleId, election })
   const precinct = getPrecinctById({ precinctId, election })
 
-  if (!precinct) {
-    throw new Error(
-      `precinct with id ${JSON.stringify(
-        precinctId
-      )} could not be found, expected one of: ${election.precincts
-        .map((p) => p.id)
-        .join(', ')}`
-    )
-  }
+  assert(ballotStyle, `invalid ballot style id: ${ballotStyleId}`)
```

Comment:
Just for my own knowledge whats the difference between doing an assert like this and doing what you do in other places where you have an if statement around ballotStyle being defined and throw an error if it isn't?

Reply:
I've been inconsistent. Both are okay. Historically I've preferred the `if/throw` style because TS could infer things based on the condition, whereas it didn't understand `assert`. In TS 3.7 they added [assertion functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions) which allow conveying this behavior. Once the `@types/node` types were updated to use this feature we can use `assert` functions to help TS do type refinement, so I've been preferring the `assert` style recently.